docs(input): better accessibility in examples

marcysutton opened the pull request:
The input and ngModel docs examples have no labels or accessible error handling. They could do more to encourage accessibility best practices. This change does two things:
1. Adds labels to each input.
2. Wraps error messages with `role="alert" so they will announce in a screen reader.

Even though Angular 2.0 is coming soon, these docs will live on with the 1.0 project so it makes sense to show more accessible examples. (Better late than never.)

Commit:
docs(): better accessibility in form examples

gkalpak commented:
I wonder if it is a good idea to use the `<label>...<input .../></label>` form (see below) in order to avoid clattering the views (and avoid having to add `id` attributes to inputs that don't need it).

```html
<label>
  Some label:
  <input type="text" name="input" ng-model="example.text" />
</label>
```

marcysutton commented:
@gkalpak because this is primarily for developer education, I opted for explicit labels on purpose. A common accessibility problem is to leave off the `for=""` and not properly link the `<label>` and the `<input>` elements (or to leave off `<label>` entirely). Showing how they are linked together does more for developer education, which has been sorely lacking in Angular in the past. 

You can read about implicit vs. explicit labeling vs. aria-labelledby here: http://www.paciellogroup.com/blog/2011/07/html5-accessibility-chops-form-control-labeling/

gkalpak commented:
Interestingly enough, although the author seems to suggest using `for / id` instead of `wrapping`, the [test results][1] seem to indicate that `wrapping` is more reliable than `for / id` ! Do I miss something ?

[1]: http://www.html5accessibility.com/tests/form-labels.html

marcysutton commented:
The support differences look pretty minimal to me. This post is also from 2012, so take it with a grain of salt.

zamith commented:
:+1:

pauljadam commented:
+1 please fix a11y in the docs it's very very bad, all the angularJS books have not a11y just like the docs, <label> wrapped labels <input> </label> are fine in my testing, yes it's less code and will look more readable, any a11y we can get the better!

ckundo commented on src/ng/directive/input.js:
```diff
@@ -68,13 +68,15 @@ var inputType = {
+           <label for="input">Single word:</label>
+           <input id="input" type="text" name="input" ng-model="example.text"
+                  ng-pattern="example.word" required ng-trim="false">
+           <span role="alert">
```
Does `role=alert` announce assertively/immediately? Is that the desired experience for a form validation?

marcysutton replied:
When I tested it, it matched the speed that the visual message would come up in context to what you're typing...it was especially useful for validations related to input length, so you'd know exactly when the error occurred.

ckundo replied:
That's really cool! Good to know.

ckundo commented on src/ng/directive/input.js:
```diff
@@ -68,13 +68,15 @@ var inputType = {
-             Single word only!</span>
-
+           <label for="input">Single word:</label>
+           <input id="input" type="text" name="input" ng-model="example.text"
```
Is it confusing to have the `id` value be the same as the element name and `name` attribute?

marcysutton replied:
Probably. I'll simplify these based on all the great feedback so far and just use implicit labeling. Thanks!

ckundo commented on src/ng/directive/input.js:
```diff
@@ -153,13 +155,15 @@ var inputType = {
+                Required!</span>
+            <span class="error" ng-show="myForm.input.$error.date">
+                Not a valid date!</span>
+           </span><br>
```
Maybe worth making this `<br>` self closing, to match other break tags in the document.

marcysutton replied:
Good point. Or it could just become a `<div>`!

ckundo commented on src/ng/directive/input.js:
```diff
@@ -153,13 +155,15 @@ var inputType = {
        </script>
        <form name="myForm" ng-controller="DateController as dateCtrl">
-          Pick a date in 2013:
+          <label for="exampleInput">Pick a date in 2013:</label>
```
It looks like `exampleInput` is a duplicate `id`. Will that impact focus and AT behavior?

marcysutton replied:
It normally would, but it's isolated to this one demo.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -943,7 +945,8 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
        <form name="testForm" ng-controller="ExampleController">
-         <input ng-model="val" ng-pattern="/^\d+$/" name="anim" class="my-input" />
+         <label for="my-input">My input</label>
+         <input id="my-input" ng-model="val" ng-pattern="/^\d+$/" name="anim" class="my-input" />
```
`id` values are camelcase elsewhere. Should that be true here too?

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1110,14 +1113,14 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
-          Name:
-          <input type="text" name="userName"
+          <label for="userName">Name:</label>
+          <input type="text" name="userName" id="userName"
```
Seems like it'd be more consistent with the document to put the `id` attribute first.

ckundo commented:
Huge +1 in general. Great to see improved accessibility in the docs, and the `role=alert` pattern on form validations is super cool and a unique accessibility feature to Angular. The duplicate `id`s might be a blocker until resolved, but otherwise, I vote ship it :rocket:!

marcysutton commented:
Thanks for the feedback! I've simplified a lot of the labeling based on comments in this PR.

Commit:
docs(): better accessibility in ng docs

marcysutton commented:
@ckundo @gkalpak I did a search for `<input>` and found a bunch more places that needed accessibility help, including stuff like this, where the "good" example is missing an alt attribute:

 > The buggy way to write it:
 ```html
 <img src="http://www.gravatar.com/avatar/{{hash}}"/>
 ```
 > The correct way to write it:
 ```html
 <img ng-src="http://www.gravatar.com/avatar/{{hash}}"/>
 ```
Can you take another look if you get a chance? Thanks so much.

ckundo commented on src/ng/directive/input.js:
```diff
@@ -68,13 +68,16 @@ var inputType = {
-
+           <label>Single word:
+             <input type="text" name="input" ng-model="example.text"
+                  ng-pattern="example.word" required ng-trim="false">
```
Alignment with the attributes above is not quite right.

ckundo commented on src/ng/directive/input.js:
```diff
@@ -433,13 +442,16 @@ var inputType = {
-        <input id="exampleInput" type="week" name="input" ng-model="example.value"
+        <label>Pick a date between in 2013:
+          <input type="week" name="input" ng-model="example.value"
             placeholder="YYYY-W##" min="2012-W32" max="2013-W52" required />
```
Attributes don't line up here (`placeholder` with `type`)

ckundo commented on src/ng/directive/input.js:
```diff
@@ -625,12 +639,16 @@ var inputType = {
-             Not valid number!</span>
+           <label>Number:
+             <input type="number" name="input" ng-model="example.value"
+                  min="0" max="99" required>
```
Attributes don't line up here.

ckundo commented on src/ng/directive/input.js:
```diff
@@ -929,9 +964,13 @@ var inputType = {
+           </label><br/>
+           <label>Value2:
+             <input type="checkbox" ng-model="checkboxModel.value2"
+                ng-true-value="'YES'" ng-false-value="'NO'">
```
Another attribute alignment issue.

ckundo commented:
Is there some guideline around line length? 80 characters is a sane guideline, but I don't know if the project adheres to that. @gkalpak do you know?

ckundo commented:
@marcysutton it occurred to me that the `role=alert` changes might be better off in a separate PR. Do you think that change can be isolated?

ckundo commented on src/ng/directive/ngClass.js:
```diff
@@ -156,17 +156,17 @@ function classDirective(name, selector) {
        <hr>
        <p ng-class="style">Using String Syntax</p>
-       <input type="text" ng-model="style" placeholder="Type: bold strike red">
+       <label><input type="text" ng-model="style" placeholder="Type: bold strike red">
```
Is this missing a closing tag?

marcysutton replied:
It is indeed! Good catch.

ckundo commented on src/ng/directive/ngClass.js:
```diff
@@ -156,17 +156,17 @@ function classDirective(name, selector) {
-       <input ng-model="style1" placeholder="Type: bold, strike or red"><br>
-       <input ng-model="style2" placeholder="Type: bold, strike or red"><br>
-       <input ng-model="style3" placeholder="Type: bold, strike or red"><br>
+       <input ng-model="style1" placeholder="Type: bold, strike or red" aria-label="Type: bold, strike or red"><br>
```
This is where a line length guideline might be nice.

ckundo replied:
Per the 100 character max guideline, this should be wrapped (right now it's 115 characters)

ckundo commented on src/ng/directive/ngController.js:
```diff
@@ -64,20 +64,20 @@
- *      Name: <input type="text" ng-model="settings.name"/>
- *      [ <a href="" ng-click="settings.greet()">greet</a> ]<br/>
+ *      <label>Name: <input type="text" ng-model="settings.name"/></label>
+ *      <button ng-click="settings.greet()">greet</button><br/>
```
Nice addition! This seems though like another instance of something that could be included in a separate PR, one related specifically to semantic use of link vs. button. Same applies for the additions below.

ckundo commented on src/ng/directive/ngController.js:
```diff
@@ -64,20 +64,20 @@
  *      <ul>
  *        <li ng-repeat="contact in settings.contacts">
- *          <select ng-model="contact.type">
+ *          <select ng-model="contact.type" aria-label="contact method" id="select_{{$index}}">
```
Does capitalization matter here for the `aria-label`? I noticed that elsewhere it's capitalized.

marcysutton replied:
It doesn't matter since it will be read aloud the same way, but I'll update it for consistency.

ckundo commented on src/ng/directive/ngController.js:
```diff
@@ -147,20 +147,20 @@
  *     <ul>
  *       <li ng-repeat="contact in contacts">
- *         <select ng-model="contact.type">
+ *         <select ng-model="contact.type" id="select_{{$index}}">
```
Cool, good call with the `aria-labelledby`! I'll defer to @gkalpak on this approach to interpolating the `$index` value, I'm not sure if there are other approaches to getting the index of the chosen option in the `select`.

gkalpak replied:
I think the `$index` approach is as good as any (as long as screen readers can cope with the id changing dynamically).

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -474,12 +474,14 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
    *       <form name="myForm" ng-model-options="{ updateOn: 'blur' }">
-   *         <p>With $rollbackViewValue()</p>
-   *         <input name="myInput1" ng-model="myValue" ng-keydown="resetWithCancel($event)"><br/>
+   *         <p id="inputLabel1">With $rollbackViewValue()</p>
```
Can't this just be a `label` instead of a `p`?

marcysutton replied:
Yeah this was a weird case where the text content didn't really work as a label for the input, but it still needed to be described by something. That's why I did it this way.

gkalpak commented:
100 characters per line is the limit.
(I believe it's somewhere in contributing.md)

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -474,12 +474,14 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
-   *         <p>With $rollbackViewValue()</p>
-   *         <input name="myInput1" ng-model="myValue" ng-keydown="resetWithCancel($event)"><br/>
+   *         <p id="inputLabel1">With $rollbackViewValue()</p>
+   *         <input name="myInput1" aria-labelledby="inputLabel1" ng-model="myValue"
```
If the `p` is in fact a description, should this be `aria-describedby`?

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -474,12 +474,14 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
    *
-   *         <p>Without $rollbackViewValue()</p>
-   *         <input name="myInput2" ng-model="myValue" ng-keydown="resetWithoutCancel($event)"><br/>
+   *         <p id="inputLabel2">Without $rollbackViewValue()</p>
```
Same question about `p` vs `label` and `labelledby` vs `describedby`

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -940,10 +942,13 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
        <form name="testForm" ng-controller="ExampleController">
-         <input ng-model="val" ng-pattern="/^\d+$/" name="anim" class="my-input" />
+         <input ng-model="val" ng-pattern="/^\d+$/" name="anim" class="my-input"
+            aria-labelledby="inputDescription" />
```
Attribute alignment is off, and same question about label association vs description.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -973,10 +978,11 @@ var NgModelController = ['$scope', '$exceptionHandler', '$attrs', '$element', '$
-           <input type="text" name="userName"
+           <label>Name:
+             <input type="text" name="userName"
                   ng-model="user.name"
```
Attribute alignment is off here.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1110,14 +1116,15 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
-          <input type="text" name="userName"
+          <label>Name:
+            <input type="text" name="userName"
                  ng-model="user.name"
```
Attribute alignment is off here.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1110,14 +1116,15 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
-          Other data:
-          <input type="text" ng-model="user.data" /><br />
+                 ng-keyup="cancel($event)" />
+          </label><br />
```
`<br>` with no closing / would be more consistent with surrounding code.

marcysutton replied:
I've seen examples with both. I can certainly make them more consistent, but it's hard to tell which convention to use :)

ckundo replied:
See Georgios comment below, looks like keeping the trailing slashes is preferred.

gkalpak replied:
Just to be clear: leaving the slashes in MY personal preference. My sense is that in most instances in the docs omitted the trailing slashes, but as mentioned already, either way is fine :)

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1110,14 +1116,15 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
+          </label><br />
+          <label>Other data:
+            <input type="text" ng-model="user.data" />
+          </label><br />
```
Sans `/` s'il vous plait

gkalpak replied:
I would happier if there were `/`s everywhere :)
<sub>(It's a matter of personal reference, so I don't think it's a big deal. The docs are full of such little inconsistencies.)</sub>

ckundo replied:
@gkalpak I agree, erring on the side of consistency with the surrounding code, to avoid complicated diffs/unrelated changes.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1165,10 +1172,11 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
-          <input type="text" name="userName"
+          <label>Name:
+            <input type="text" name="userName"
                  ng-model="user.name"
```
Attribute alignment.

ckundo commented on src/ng/directive/ngModel.js:
```diff
@@ -1188,10 +1196,11 @@ var DEFAULT_REGEXP = /(\s+|^)default(\s+|$)/;
-          <input type="text" name="userName"
+          <label>Name:
+            <input type="text" name="userName"
                  ng-model="user.name"
```
Attribute alignment

gkalpak commented on src/ng/directive/attrs.js:
````diff
@@ -144,7 +144,7 @@
  * The correct way to write it:
  * ```html
- * <img ng-srcset="http://www.gravatar.com/avatar/{{hash}} 2x"/>
+ * <img ng-srcset="http://www.gravatar.com/avatar/{{hash}} 2x" alt="Description" />
````
I suggest the same change is made to the buggy way above.
Since this example is about `src` vs `ngSrc`, we don't want to distract the reader into thinking `alt` has anything to do with that.

marcysutton replied:
Gotcha. That makes sense to me.

ckundo replied:
@gkalpak @marcysutton why not add the `alt` to both?

marcysutton replied:
Done! Great suggestion.

gkalpak commented on src/ng/directive/attrs.js:
````diff
@@ -118,7 +118,7 @@
  * The correct way to write it:
  * ```html
- * <img ng-src="http://www.gravatar.com/avatar/{{hash}}"/>
+ * <img ng-src="http://www.gravatar.com/avatar/{{hash}}" alt="Description" />
````
See https://github.com/angular/angular.js/pull/11079/files?diff=unified#r25614911.

ckundo commented on src/ng/directive/ngOptions.js:
```diff
@@ -147,37 +147,39 @@ var ngOptionsMinErr = minErr('ngOptions');
-              [<a href ng-click="colors.splice($index, 1)">X</a>]
+              <label>Name: <input ng-model="color.name"></label>
+              <label><input type="checkbox" ng-model="color.notAnOption"> Disabled?</label>
+              [<button ng-click="colors.splice($index, 1)" aria-label="remove">X</button>]
```
Maybe a separate PR. Also, in a similar change earlier in these changes the brackets were dropped. That's inconsistent with this change. Not sure which you prefer.

marcysutton replied:
Good catch. Without brackets works best with buttons.

ckundo commented on src/ng/directive/ngOptions.js:
```diff
@@ -147,37 +147,39 @@ var ngOptionsMinErr = minErr('ngOptions');
-          Color (null allowed):
+          <label>Color (null not allowed):
+            <select ng-model="myColor" ng-options="color.name for color in colors"></select><br>
+          </label>
```
Something is wonky with this whitespace.

ckundo commented on src/ng/directive/ngOptions.js:
```diff
@@ -147,37 +147,39 @@ var ngOptionsMinErr = minErr('ngOptions');
-          </select><br/>
+          <label>Color grouped by shade, with some disabled:
+            <select ng-model="myColor"
+                    ng-options="color.name group by color.shade disable when color.notAnOption for color in colors">
```
This looks like it might be another line length concern.

marcysutton replied:
Can we break the `ng-options` value string up at all? I think it was over the line length already before I made the edit, but if there's a way to make it fit better, I'll do it!

gkalpak replied:
At least in Chrome, you can break strings up with no problem. But since this is one long string I believe it's OK even if it's a little longer than 100 chars.

ckundo commented on src/ng/directive/ngShowHide.js:
```diff
@@ -98,7 +98,7 @@ var NG_HIDE_IN_PROGRESS_CLASS = 'ng-hide-animate';
   <example module="ngAnimate" deps="angular-animate.js" animations="true">
     <file name="index.html">
-      Click me: <input type="checkbox" ng-model="checked"><br/>
+      Click me: <input type="checkbox" ng-model="checked" aria-label="toggle ngHide and ngShow"><br/>
```
Is this actually what's happening ("toggle ngHide and ngShow")? It seems like this is actually toggling `.ng-hide`. @gkalpak can you weigh in?

gkalpak replied:
Toggling `.ng-hide` is an implementation detail that is of no interest to the user.
Maybe "toggle thumbs-up thumbs-down icons" ?

<sub>(Again, this is not a real app, so it isn't all that significant. The important thing here is to demonstrate proper use of aria attributes.)</sub>

ckundo commented on src/ng/directive/ngShowHide.js:
```diff
@@ -263,7 +263,7 @@ var ngShowDirective = ['$animate', function($animate) {
   <example module="ngAnimate" deps="angular-animate.js" animations="true">
     <file name="index.html">
-      Click me: <input type="checkbox" ng-model="checked"><br/>
+      Click me: <input type="checkbox" ng-model="checked" aria-label="toggle ngShow and ngHide"><br/>
```
Same comment, what are we actually toggling here?

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -879,9 +905,18 @@ var inputType = {
+           <label>
+             <input id="blue" type="radio" ng-model="color.name" value="blue">
+             Blue
+           </label><br/>
```
Why add ids, if are nesting them into labels ?

marcysutton replied:
Good question, that was leftover from my first pass on the code. They are no longer necessary!

ckundo commented on src/ng/filter/limitTo.js:
```diff
@@ -37,11 +37,11 @@
+         <label>Limit {{letters}} to: <input type="number" step="1" ng-model="letterLimit"></label>
          <p>Output letters: {{ letters | limitTo:letterLimit }}</p>
-         Limit {{longNumber}} to: <input type="number" step="1" ng-model="longNumberLimit">
+         <label>Limit {{longNumber}} to: <input type="number" step="1" ng-model="longNumberLimit"></label>
```
Line length?

ckundo commented on src/ng/http.js:
```diff
@@ -671,11 +671,11 @@ function $HttpProvider() {
 <file name="index.html">
   <div ng-controller="FetchController">
-    <select ng-model="method">
+    <select ng-model="method" aria-label="HTTP method">
```
I don't think JSONP is an HTTP method, maybe the label should be "Request method"

gkalpak replied:
:+1: Good catch

gkalpak commented on src/ng/directive/input.js:
```diff
@@ -1449,15 +1488,23 @@ function checkboxInputType(scope, element, attr, ctrl, $sniffer, $browser, $filt
+           </div>
+           <label>Last name:
+             <input type="text" name="lastName" ng-model="user.last"
              ng-minlength="3" ng-maxlength="10">
```
This seems misaligned...

ckundo commented on src/ng/sce.js:
````diff
@@ -413,7 +413,10 @@ function $SceDelegateProvider() {
  * ```
- * <input ng-model="userHtml">
+ * <label>
+ *   User input
````
This seems to be changing the UI by adding additional content. Seems like that should be fine, but thought I'd call it out since so far that has not been the case.

marcysutton replied:
I will switch it to `aria-label` so that we don't have to add new content.

ckundo commented on src/ngMessages/messages.js:
````diff
@@ -31,7 +31,9 @@ var jqLite = angular.element;
  * ```html
  * <form name="myForm">
- *   <input type="text" ng-model="field" name="myField" required minlength="5" />
+ *   <label>Enter text:
````
Same question, is adding this new label text in scope?

marcysutton replied:
This is pretty basic form label stuff, in this kind of an example to me it makes sense to add label text.

gkalpak commented on src/ng/directive/ngList.js:
```diff
@@ -30,9 +30,11 @@
+ *      <label>List: <input name="namesInput" ng-model="names" ng-list required></label>
+ *      <span role="alert">
+ *        <span class="error" ng-show="myForm.namesInput.$error.required">
  *        Required!</span>
```
Indentation...

gkalpak commented on src/ngMessages/messages.js:
```diff
@@ -31,7 +31,9 @@ var jqLite = angular.element;
+ *   <label>Enter text:
+ *     <input type="text" ng-model="field" name="myField" required minlength="5" />
+ *   </label>
  *   <div ng-messages="myForm.myField.$error">
```
Could this use a `role="alert"` attrbute ?

marcysutton replied:
`ngAria` actually adds them to `ngMessages`, but it can't hurt to include the role here.

ckundo commented:
Besides the formatting fixes, I think the `role=alert` and `a` vs `button` concerns might be better in a separate PR, but that may be a little fussy. Also noted that the last file in the changes has new content, which was different. All the other changes should have no effect on the presentation of the page, so thought I'd raise it.

@marcysutton thanks a million for this work. Keep it up!

gkalpak commented:
It generally LGTM. Really great work ! :+1: @marcysutton 

--
A couple of notes:

1. Alignment is not super-correct. It would nice to have everything align/indented consistently of course, but I can leave without as well :)

2. I prefer closing `/>` (**with** the `/`), but that seems to not be directly related to this PR (plus is a matter of personal preference anyway).

3. Although I generally agree that `button` is preferable to `a` (when used as a button), `a`s do look better in certain cases. Couldn't we go with the `a`s (e.g. for `X` "buttons") and put a `role` attribute or something. Since many people do use `a`s anyway, it would be good to raise awareness around "doing it right" (if you have to).

marcysutton commented:
Thanks again for your thorough review, this is awesome!

1. Edited alignment to be more consistent.

2. I did a find and replace for `<br>` in the files I had edited and changed any occurrences to `<br/>` since it was pretty easy.

3. The reason I changed the `<a>`s to `<button>`s is because they all had empty `href`s and were essentially doing the work of buttons (ng-click only, no routing). Adding `role=button` wouldn't negate the need for an empty `href` to make it a valid element. It's just easier to make them buttons. :)

If it's okay to leave this all in one branch, it will save me from having to do the work of splitting it up and creating additional PRs. In the future I'll try to break it up though. I know it's a big one! Thanks again for all your help.

ckundo commented:
@marcysutton thanks, LGTM! :+1:

marcysutton commented on src/ng/compile.js:
```diff
@@ -621,8 +621,8 @@
-      <input ng-model="name"> <br>
-      <textarea ng-model="html"></textarea> <br>
+      <input ng-model="name"> <br/>
+      <textarea ng-model="html"></textarea> <br/>
```
These should have labels too. Oops! I'll fix them up later.
